File: app/src/main/java/org/nevack/unitconverter/model/converter/TemperatureConverter.java
package org.nevack.unitconverter.model.converter;

import android.content.Context;

import org.nevack.unitconverter.R;
import org.nevack.unitconverter.model.Unit;

import java.math.BigDecimal;

public class TemperatureConverter extends Converter {
    public TemperatureConverter(Context context) {
        units.add(new Unit(context.getString(R.string.kelvin), 1d, context.getString(R.string.kelvinsymbol)));
        units.add(new Unit(context.getString(R.string.celsius), 1d, context.getString(R.string.celsiussymbol)));
        units.add(new Unit(context.getString(R.string.fahrenheit), 1d, context.getString(R.string.fahrenheitsymbol)));
    }

    @Override
    public String convert(String inputValue, int inputValueType, int outputValueType) {

        if (inputValueType == outputValueType) return inputValue;

        BigDecimal sourceValue = new BigDecimal(inputValue);
        String resultValue;

        switch (inputValueType) {
            case 0:
                if (sourceValue.compareTo(BigDecimal.ZERO) < 0) throw new ArithmeticException();
                else {
                    if (outputValueType == 1) {
                        resultValue = sourceValue.add(new BigDecimal("-273"))
                                .stripTrailingZeros()
                                .toPlainString();
                    } else {
                        resultValue = sourceValue.multiply(new BigDecimal("1.8"))
                                .add(new BigDecimal("-459.67"))
                                .stripTrailingZeros()
                                .toPlainString();
                    }
                }
                break;
            case 1:
                if (sourceValue.compareTo(new BigDecimal("-273")) < 0) throw new ArithmeticException();
                else {
                    if (outputValueType == 0) {
                        resultValue = sourceValue.add(new BigDecimal("273"))
                                .stripTrailingZeros()
                                .toPlainString();
                    } else {
                        resultValue = sourceValue.multiply(new BigDecimal("1.8"))
                                .add(new BigDecimal("32"))
                                .stripTrailingZeros()
                                .toPlainString();
                    }
                }
                break;
            case 2:
                if (sourceValue.compareTo(new BigDecimal("-459.67")) < 0) throw new ArithmeticException();
                else {
                    if (outputValueType == 0) {
                        resultValue = sourceValue.add(new BigDecimal("459.67"))
                                .multiply(new BigDecimal("5"))
                                .divide(new BigDecimal("9"), SCALE, BigDecimal.ROUND_HALF_UP)
                                .stripTrailingZeros()
                                .toPlainString();
                    } else {
                        resultValue = sourceValue.add(new BigDecimal("-32"))
                                .divide(new BigDecimal("1.8"), SCALE, BigDecimal.ROUND_HALF_UP)
                                .stripTrailingZeros()
                                .toPlainString();
                    }
                }
                break;
            default:
                throw new ArithmeticException();
        }

        return resultValue;
    }

    public int getName() {
        return R.string.temperature;
    }
}

File: app/src/main/java/org/nevack/unitconverter/model/converter/TemperatureConverter.kt
package org.nevack.unitconverter.model.converter

import android.content.Context
import java.lang.ArithmeticException
import org.nevack.unitconverter.R
import org.nevack.unitconverter.model.ConversionUnit
import java.math.BigDecimal

class TemperatureConverter(private val context: Context) : Converter(R.string.temperature) {
    override suspend fun load() {
        registerUnit(
            ConversionUnit(
                context.getString(R.string.kelvin),
                1.0,
                context.getString(R.string.kelvinsymbol)
            )
        )
        registerUnit(
            ConversionUnit(
                context.getString(R.string.celsius),
                1.0,
                context.getString(R.string.celsiussymbol)
            )
        )
        registerUnit(
            ConversionUnit(
                context.getString(R.string.fahrenheit),
                1.0,
                context.getString(R.string.fahrenheitsymbol)
            )
        )
    }

    @Throws(ArithmeticException::class)
    override fun convert(inputValue: String, inputValueType: Int, outputValueType: Int): String {
        val sourceValue = BigDecimal(inputValue, MC)
        return inputValueType.temperatureUnit()
            .convertTo(outputValueType.temperatureUnit(), sourceValue)
            .stripTrailingZeros()
            .toPlainString()
    }

    private fun Int.temperatureUnit(): TemperatureUnit = when (this) {
        TemperatureUnit.Kelvin.index -> TemperatureUnit.Kelvin
        TemperatureUnit.Celsius.index -> TemperatureUnit.Celsius
        TemperatureUnit.Fahrenheit.index -> TemperatureUnit.Fahrenheit
        else -> throw ArithmeticException()
    }

    private sealed class TemperatureUnit(
        val index: Int,
        val convertTo: (TemperatureUnit, BigDecimal) -> BigDecimal,
    ) {
        object Kelvin : TemperatureUnit(0, { unit, x ->
            when (unit) {
                Kelvin -> x
                Celsius -> x - MIN_KELVIN
                Fahrenheit -> Celsius.convertTo(Fahrenheit, x - MIN_KELVIN)
            }
        })

        object Celsius : TemperatureUnit(1, { unit, x ->
            when (unit) {
                Kelvin -> x + MIN_KELVIN
                Celsius -> x
                Fahrenheit -> x * FAHRENHEIT_MULTIPLIER + FAHRENHEIT_OFFSET
            }
        })

        object Fahrenheit : TemperatureUnit(2, { unit, x ->
            when (unit) {
                Kelvin -> Fahrenheit.convertTo(Celsius, x) + MIN_KELVIN
                Celsius -> (x - FAHRENHEIT_OFFSET) / FAHRENHEIT_MULTIPLIER
                Fahrenheit -> x
            }
        })

        companion object {
            private val MIN_KELVIN = BigDecimal("-273.15", MC)
            private val FAHRENHEIT_OFFSET = BigDecimal(32, MC)
            private val FAHRENHEIT_MULTIPLIER = BigDecimal(9, MC) / BigDecimal(5, MC)
        }
    }
}